### PR TITLE
win,tty: Use ANSI and Xterm processing of ConEmu

### PR DESCRIPTION
**Pull request by erw7**
ConEmu has its own [ANSI X3.64 and and its extension xterm 256 color mode processing](https://conemu.github.io/en/AnsiEscapeCodes.html). But, currently libuv does not use it.

I have created a patch that uses it in libuv to improve TUI of neovim on Windows. If there is no problem, I would like you to merge it upstream. 

Ref neovim/neovim#8462

### CODE REVIEW

File: src/win/tty.c [hunk 1]
```diff
@@ -2259,25 +2262,95 @@ int uv_tty_reset_mode(void) {
...
+                                   FALSE,
+                                   pbi.InheritedFromUniqueProcessId);
+      if (!process_handle) {
+        break;
+      }
+
+      parent_file_name_length =
+        GetProcessImageFileNameW(process_handle,
+                                 parent_file_name,
+                                 sizeof(parent_file_name));
```
Comment (piscisaureus) [inline src/win/tty.c]: Size should be in characters, not bytes. This could lead to writing past the end of the buffer.

**Comment by bzoz** [comment]
Is this needed? It adds complexity and maintenance burden. It also looks like emulation works on ConEmu just fine.

**Comment by bnoordhuis** [comment]
I think this is for features that conemu supports but regular console windows don't.  It might be reasonable to support in libuv if it's not too much hassle.

That said, adding a Windows-only `uv_guess_tty()` is out of the question and I don't know if I'd say that a 200+ line diff qualifies as 'not too much hassle.'  I defer to our Windows people.

**Comment by erw7** [comment]
@bzoz  Certainly emulation of libuv works fine with ConEmu. But emulation of ConEmu can handle 256 colors and 24 bit colors. I need this feature because it becomes possible to use `termguicolors` if neovim can use it. 8 colors and 256 colors or 24 bit colors are a big difference in neovim.

If this is too complicated, I think that it may be okay just to check the environment variables(`ConEmuANSI`).

File: src/win/tty.c [hunk 2]
```diff
@@ -2259,25 +2264,95 @@ int uv_tty_reset_mode(void) {
...
+        WCHAR* comp_position;
+        if (parent_file_name_length < conemu_file_name_length) {
+          continue;
+        }
+        comp_position = parent_file_name + parent_file_name_length
+          - conemu_file_name_length;
+        if (!wcsncmp(comp_position,
+                     conemu_file_names[i],
+                     conemu_file_name_length)) {
+          result |= UV_TTY_CONEMU;
```
[the remark this answers is not in the document]
Comment (erw7) [reply src/win/tty.c]: If we use only the function of ConEmu(ANSI X3.64 and Xterm 256 colors), I think that it is enough to check `ConEmu` environment variables. However If we try to deal with when running on `winpty`, this code is absolutely necessary. I think that in order to resolve neovim/neovim#9016 we need to do so.

**Comment by bzoz** [comment]
@justinmk: that would probably be the best solution, I'm +1 on that.

**Comment by erw7** [comment]
I changed `uv_set_vterm_state` of #1884 from private to public. Therefore, I close this pull request.

**Comment by justinmk** [comment]
@erw7 could you separate-out the low-risk changes from #1884 into a separate PR? For example, exposing `uv_set_vterm_state` is low-risk and could be merged more quickly than the bigger changes in #1884.

**Comment by erw7** [comment]
@justinmk I call `uv_set_vterm_state` in `test-tty-escape-sequence-processing.c`. Therefore, exposing `uv_set_vterm_state` is necessary for #1884. It is possible to split if the same change does not cause problems even if it exists in multiple pull requests.

cc @piscisaureus As justinmk says, is it better to separate  #1844?

**Comment by vtjnash** [comment]
Yes, if you can make a new PR to add `uv_set_vterm_state`, I think we can get that merged fairly quickly, then move on to getting the rest of #1884 merged.

One pre-emptive comments for the PR:
It be helpful if toggling the `uv_set_vterm_state` would also update the console state appropriately too (similar to changing the mode). I think that means it should support the options provided here, not just the internal ones being exposed in #1884:
```
typedef enum {
  UV_TTY_AUTODETECT, /* unchecked */
  UV_TTY_LEGACY,
  UV_TTY_VTP, /* modern conhost */
  UV_TTY_ANSI /* conemu, mintty, etc. */
} uv_vtermstate_t;
```

**Pull request closed without merging**